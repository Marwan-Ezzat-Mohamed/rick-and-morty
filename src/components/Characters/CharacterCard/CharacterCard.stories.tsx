import React from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';

import CharacterCard from '.';
import { mockCharacters } from '../../../__mocks__/characters.mock';

export default {
  title: 'Components/Character Card',
  component: CharacterCard,
} as ComponentMeta<typeof CharacterCard>;

const Template: ComponentStory<typeof CharacterCard> = args => (
  <CharacterCard {...args} />
);

export const Primary = Template.bind({});
Primary.args = {
  character: mockCharacters.characters.results[0],
};

export const Skeleton = Template.bind({});
Skeleton.args = {};
