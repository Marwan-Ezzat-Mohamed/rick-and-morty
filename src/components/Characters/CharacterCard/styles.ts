import { SxProps } from '@mui/material';

const CharacterStyles: Record<string, SxProps> = {
  card: {
    width: '255px',
    height: '495px',
    p: 2,
    borderRadius: '20px',
    boxShadow: 6,
    display: 'flex',
    flexDirection: 'column',
  },
  media: {
    height: '215px',
    borderRadius: '0.7rem',
    boxShadow: 5,
    cursor: 'pointer',
  },
  content: {
    p: 0,
    paddingY: '0px !important',
    mt: 2,
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'space-between',
    flexGrow: 1,
  },
  status: {
    height: '0.65rem',
    width: '0.65rem',
    borderRadius: '50%',
  },
  statusContainer: {
    display: 'flex',
    alignItems: 'center',
    p: 0,
  },

  infoText: {
    lineHeight: '1.3',
    overflow: 'hidden',
    textOverflow: 'ellipsis',
    '-webkit-line-clamp': '2',
    '-webkit-box-orient': 'vertical',
    display: '-webkit-box',
    //hover
    '&:hover': {
      overflowY: 'auto',
      textOverflow: 'unset',
    },
  },
};

export default CharacterStyles;
