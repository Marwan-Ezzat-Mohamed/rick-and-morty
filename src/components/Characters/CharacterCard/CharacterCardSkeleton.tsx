import Card from '@mui/material/Card';
import CardContent from '@mui/material/CardContent';
import Skeleton from '@mui/material/Skeleton';
import Typography from '@mui/material/Typography';
import Box from '@mui/material/Box';
import { useTheme } from '@mui/material/styles';
import { tokens } from './../../../theme';
import styles from './styles';
import { memo } from 'react';

const CharacterCardSkeleton = () => {
  const theme = useTheme();
  const colors = tokens(theme.palette.mode);

  const Information = () => {
    return (
      <Box>
        <Box
          sx={{
            display: 'flex',
            color: '#8f949f',
            alignItems: 'center',
          }}
        >
          <Typography variant="h5" width="40%">
            <Skeleton />
          </Typography>
        </Box>
        <Typography variant="h5">
          <Skeleton />
        </Typography>
      </Box>
    );
  };

  return (
    <Card
      sx={{
        ...styles.card,
        backgroundColor: colors.primary[400],
        '&:hover': {
          backgroundColor: colors.primary[500],
          transform: 'scale(1.05)',
        },
      }}
    >
      <Skeleton variant="rectangular" sx={styles.media} />
      <CardContent sx={styles.content}>
        <Typography variant="h3" fontWeight="bold">
          <Skeleton />
        </Typography>

        {Information()}
        {Information()}
        {Information()}
      </CardContent>
    </Card>
  );
};

export default memo(CharacterCardSkeleton);
