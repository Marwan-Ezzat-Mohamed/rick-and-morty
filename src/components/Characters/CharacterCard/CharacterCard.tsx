import React, { useMemo, memo } from 'react';
import Card from '@mui/material/Card';
import CardContent from '@mui/material/CardContent';
import CardMedia from '@mui/material/CardMedia';

import Typography from '@mui/material/Typography';
import { useTheme } from '@mui/material/styles';
import styles from './styles';
import { getTokens } from '../../../Theme/theme';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import {
  faDna,
  faLocationArrow,
  faEye,
  IconDefinition,
} from '@fortawesome/free-solid-svg-icons';
import Box from '@mui/material/Box';
import { Character } from '../../../GraphQL/types';
import CharacterCardSkeleton from './CharacterCardSkeleton';
import shallow from 'zustand/shallow';

export interface ICharacterCardProps {
  character: Character | null;
}

const CharacterCard = ({ character }: ICharacterCardProps) => {
  const theme = useTheme();
  const colors = getTokens(theme.palette.mode);
  const characterStatus: Record<string, string> = useMemo(
    () => ({
      Alive: colors.greenAccent[500],
      Dead: colors.redAccent[500],
      unknown: colors.grey[500],
    }),
    [colors],
  );
  if (!character) return <CharacterCardSkeleton />;

  const Information = ({
    title,
    value,
    icon,
    rotate = '0deg',
  }: {
    title: string;
    value: string;
    icon: IconDefinition;
    rotate?: string;
  }) => {
    return (
      <Box>
        <Box
          sx={{
            display: 'flex',
            color: '#8f949f',
            alignItems: 'center',
          }}
        >
          <FontAwesomeIcon
            icon={icon}
            style={{
              rotate: rotate,
              fontWeight: 'bold',
            }}
          />
          <Typography variant="h5" component="div" ml={1} fontWeight="bold">
            {title}
          </Typography>
        </Box>
        <Typography variant="h5" component="div" sx={styles.infoText}>
          {value}
        </Typography>
      </Box>
    );
  };

  const { image, name, status, species, gender, location, episode } = character;

  return (
    <Card
      sx={{
        ...styles.card,
        backgroundColor: colors.primary[400],
        '&:hover': {
          backgroundColor: colors.primary[500],
          transform: 'scale(1.04)',
        },
      }}
    >
      <CardMedia component="img" sx={styles.media} image={image} alt={name} />
      <CardContent sx={styles.content}>
        <Typography variant="h3" component="div" fontWeight="bold">
          {name}
        </Typography>

        <Box sx={styles.statusContainer}>
          <Box
            sx={{
              ...styles.status,
              backgroundColor: characterStatus[status],
              boxShadow:
                status !== 'unknown'
                  ? `0rem 0rem 0.45rem 0.04rem ${characterStatus[status]}`
                  : 'none',
            }}
          />
          <Typography variant="h5" component="div" ml={1}>
            {`${status}`}
          </Typography>
        </Box>

        <Information
          title="Species"
          value={`${species} - ${gender}`}
          rotate="45deg"
          icon={faDna}
        />
        <Information
          title="Last known location"
          value={location.name}
          rotate="45deg"
          icon={faLocationArrow}
        />
        <Information
          title="First seen in"
          value={episode[0].name}
          icon={faEye}
        />
      </CardContent>
    </Card>
  );
};

export default memo(CharacterCard, (prevProps, nextProps) => {
  return shallow(prevProps.character, nextProps.character);
});
