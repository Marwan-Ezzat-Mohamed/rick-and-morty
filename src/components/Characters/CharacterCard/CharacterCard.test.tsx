import '@testing-library/jest-dom';
import { render } from '@testing-library/react';
import CharacterCard from './CharacterCard';
import { mockCharacters } from './../../../__mocks__/characters.mock';

const character = mockCharacters.characters.results[0];

describe('CharacterCard', () => {
  it('matches snapshot', () => {
    const rendered = render(<CharacterCard character={character} />);
    expect(rendered).toMatchSnapshot();
  });

  it('renders skeleton when character is null', () => {
    const rendered = render(<CharacterCard character={null} />);
    expect(
      rendered.container.querySelector('.MuiSkeleton-root'),
    ).toBeInTheDocument();
  });
});
