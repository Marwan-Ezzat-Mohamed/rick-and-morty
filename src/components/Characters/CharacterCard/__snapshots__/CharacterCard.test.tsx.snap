// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterCard matches snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1amsolq-MuiPaper-root-MuiCard-root"
      >
        <img
          alt="Rick Sanchez"
          class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-p178uf-MuiCardMedia-root"
          src="https://rickandmortyapi.com/api/character/avatar/1.jpeg"
        />
        <div
          class="MuiCardContent-root css-1wkgrvo-MuiCardContent-root"
        >
          <div
            class="MuiTypography-root MuiTypography-h3 css-1ey9xtw-MuiTypography-root"
          >
            Rick Sanchez
          </div>
          <div
            class="MuiBox-root css-s4ncqe"
          >
            <div
              class="MuiBox-root css-14qgaio"
            />
            <div
              class="MuiTypography-root MuiTypography-h5 css-pk4xan-MuiTypography-root"
            >
              Alive
            </div>
          </div>
          <div
            class="MuiBox-root css-0"
          >
            <div
              class="MuiBox-root css-11rx8v9"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-dna "
                data-icon="dna"
                data-prefix="fas"
                focusable="false"
                role="img"
                style="font-weight: bold;"
                viewBox="0 0 448 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M416 0c17.7 0 32 14.3 32 32c0 59.8-30.3 107.5-69.4 146.6c-28 28-62.5 53.5-97.3 77.4l-2.5 1.7c-11.9 8.1-23.8 16.1-35.5 23.9l0 0 0 0 0 0-1.6 1c-6 4-11.9 7.9-17.8 11.9c-20.9 14-40.8 27.7-59.3 41.5H283.3c-9.8-7.4-20.1-14.7-30.7-22.1l7-4.7 3-2c15.1-10.1 30.9-20.6 46.7-31.6c25 18.1 48.9 37.3 69.4 57.7C417.7 372.5 448 420.2 448 480c0 17.7-14.3 32-32 32s-32-14.3-32-32H64c0 17.7-14.3 32-32 32s-32-14.3-32-32c0-59.8 30.3-107.5 69.4-146.6c28-28 62.5-53.5 97.3-77.4c-34.8-23.9-69.3-49.3-97.3-77.4C30.3 139.5 0 91.8 0 32C0 14.3 14.3 0 32 0S64 14.3 64 32H384c0-17.7 14.3-32 32-32zM338.6 384H109.4c-10.1 10.6-18.6 21.3-25.5 32H364.1c-6.8-10.7-15.3-21.4-25.5-32zM109.4 128H338.6c10.1-10.7 18.6-21.3 25.5-32H83.9c6.8 10.7 15.3 21.3 25.5 32zm55.4 48c18.4 13.8 38.4 27.5 59.3 41.5c20.9-14 40.8-27.7 59.3-41.5H164.7z"
                  fill="currentColor"
                />
              </svg>
              <div
                class="MuiTypography-root MuiTypography-h5 css-hi164r-MuiTypography-root"
              >
                Species
              </div>
            </div>
            <div
              class="MuiTypography-root MuiTypography-h5 css-18vl5hq-MuiTypography-root"
            >
              Human - Male
            </div>
          </div>
          <div
            class="MuiBox-root css-0"
          >
            <div
              class="MuiBox-root css-11rx8v9"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-location-arrow "
                data-icon="location-arrow"
                data-prefix="fas"
                focusable="false"
                role="img"
                style="font-weight: bold;"
                viewBox="0 0 448 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M429.6 92.1c4.9-11.9 2.1-25.6-7-34.7s-22.8-11.9-34.7-7l-352 144c-14.2 5.8-22.2 20.8-19.3 35.8s16.1 25.8 31.4 25.8H224V432c0 15.3 10.8 28.4 25.8 31.4s30-5.1 35.8-19.3l144-352z"
                  fill="currentColor"
                />
              </svg>
              <div
                class="MuiTypography-root MuiTypography-h5 css-hi164r-MuiTypography-root"
              >
                Last known location
              </div>
            </div>
            <div
              class="MuiTypography-root MuiTypography-h5 css-18vl5hq-MuiTypography-root"
            >
              Citadel of Ricks
            </div>
          </div>
          <div
            class="MuiBox-root css-0"
          >
            <div
              class="MuiBox-root css-11rx8v9"
            >
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-eye "
                data-icon="eye"
                data-prefix="fas"
                focusable="false"
                role="img"
                style="font-weight: bold;"
                viewBox="0 0 576 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M288 32c-80.8 0-145.5 36.8-192.6 80.6C48.6 156 17.3 208 2.5 243.7c-3.3 7.9-3.3 16.7 0 24.6C17.3 304 48.6 356 95.4 399.4C142.5 443.2 207.2 480 288 480s145.5-36.8 192.6-80.6c46.8-43.5 78.1-95.4 93-131.1c3.3-7.9 3.3-16.7 0-24.6c-14.9-35.7-46.2-87.7-93-131.1C433.5 68.8 368.8 32 288 32zM432 256c0 79.5-64.5 144-144 144s-144-64.5-144-144s64.5-144 144-144s144 64.5 144 144zM288 192c0 35.3-28.7 64-64 64c-11.5 0-22.3-3-31.6-8.4c-.2 2.8-.4 5.5-.4 8.4c0 53 43 96 96 96s96-43 96-96s-43-96-96-96c-2.8 0-5.6 .1-8.4 .4c5.3 9.3 8.4 20.1 8.4 31.6z"
                  fill="currentColor"
                />
              </svg>
              <div
                class="MuiTypography-root MuiTypography-h5 css-hi164r-MuiTypography-root"
              >
                First seen in
              </div>
            </div>
            <div
              class="MuiTypography-root MuiTypography-h5 css-18vl5hq-MuiTypography-root"
            >
              Pilot
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1amsolq-MuiPaper-root-MuiCard-root"
    >
      <img
        alt="Rick Sanchez"
        class="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-p178uf-MuiCardMedia-root"
        src="https://rickandmortyapi.com/api/character/avatar/1.jpeg"
      />
      <div
        class="MuiCardContent-root css-1wkgrvo-MuiCardContent-root"
      >
        <div
          class="MuiTypography-root MuiTypography-h3 css-1ey9xtw-MuiTypography-root"
        >
          Rick Sanchez
        </div>
        <div
          class="MuiBox-root css-s4ncqe"
        >
          <div
            class="MuiBox-root css-14qgaio"
          />
          <div
            class="MuiTypography-root MuiTypography-h5 css-pk4xan-MuiTypography-root"
          >
            Alive
          </div>
        </div>
        <div
          class="MuiBox-root css-0"
        >
          <div
            class="MuiBox-root css-11rx8v9"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-dna "
              data-icon="dna"
              data-prefix="fas"
              focusable="false"
              role="img"
              style="font-weight: bold;"
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M416 0c17.7 0 32 14.3 32 32c0 59.8-30.3 107.5-69.4 146.6c-28 28-62.5 53.5-97.3 77.4l-2.5 1.7c-11.9 8.1-23.8 16.1-35.5 23.9l0 0 0 0 0 0-1.6 1c-6 4-11.9 7.9-17.8 11.9c-20.9 14-40.8 27.7-59.3 41.5H283.3c-9.8-7.4-20.1-14.7-30.7-22.1l7-4.7 3-2c15.1-10.1 30.9-20.6 46.7-31.6c25 18.1 48.9 37.3 69.4 57.7C417.7 372.5 448 420.2 448 480c0 17.7-14.3 32-32 32s-32-14.3-32-32H64c0 17.7-14.3 32-32 32s-32-14.3-32-32c0-59.8 30.3-107.5 69.4-146.6c28-28 62.5-53.5 97.3-77.4c-34.8-23.9-69.3-49.3-97.3-77.4C30.3 139.5 0 91.8 0 32C0 14.3 14.3 0 32 0S64 14.3 64 32H384c0-17.7 14.3-32 32-32zM338.6 384H109.4c-10.1 10.6-18.6 21.3-25.5 32H364.1c-6.8-10.7-15.3-21.4-25.5-32zM109.4 128H338.6c10.1-10.7 18.6-21.3 25.5-32H83.9c6.8 10.7 15.3 21.3 25.5 32zm55.4 48c18.4 13.8 38.4 27.5 59.3 41.5c20.9-14 40.8-27.7 59.3-41.5H164.7z"
                fill="currentColor"
              />
            </svg>
            <div
              class="MuiTypography-root MuiTypography-h5 css-hi164r-MuiTypography-root"
            >
              Species
            </div>
          </div>
          <div
            class="MuiTypography-root MuiTypography-h5 css-18vl5hq-MuiTypography-root"
          >
            Human - Male
          </div>
        </div>
        <div
          class="MuiBox-root css-0"
        >
          <div
            class="MuiBox-root css-11rx8v9"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-location-arrow "
              data-icon="location-arrow"
              data-prefix="fas"
              focusable="false"
              role="img"
              style="font-weight: bold;"
              viewBox="0 0 448 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M429.6 92.1c4.9-11.9 2.1-25.6-7-34.7s-22.8-11.9-34.7-7l-352 144c-14.2 5.8-22.2 20.8-19.3 35.8s16.1 25.8 31.4 25.8H224V432c0 15.3 10.8 28.4 25.8 31.4s30-5.1 35.8-19.3l144-352z"
                fill="currentColor"
              />
            </svg>
            <div
              class="MuiTypography-root MuiTypography-h5 css-hi164r-MuiTypography-root"
            >
              Last known location
            </div>
          </div>
          <div
            class="MuiTypography-root MuiTypography-h5 css-18vl5hq-MuiTypography-root"
          >
            Citadel of Ricks
          </div>
        </div>
        <div
          class="MuiBox-root css-0"
        >
          <div
            class="MuiBox-root css-11rx8v9"
          >
            <svg
              aria-hidden="true"
              class="svg-inline--fa fa-eye "
              data-icon="eye"
              data-prefix="fas"
              focusable="false"
              role="img"
              style="font-weight: bold;"
              viewBox="0 0 576 512"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M288 32c-80.8 0-145.5 36.8-192.6 80.6C48.6 156 17.3 208 2.5 243.7c-3.3 7.9-3.3 16.7 0 24.6C17.3 304 48.6 356 95.4 399.4C142.5 443.2 207.2 480 288 480s145.5-36.8 192.6-80.6c46.8-43.5 78.1-95.4 93-131.1c3.3-7.9 3.3-16.7 0-24.6c-14.9-35.7-46.2-87.7-93-131.1C433.5 68.8 368.8 32 288 32zM432 256c0 79.5-64.5 144-144 144s-144-64.5-144-144s64.5-144 144-144s144 64.5 144 144zM288 192c0 35.3-28.7 64-64 64c-11.5 0-22.3-3-31.6-8.4c-.2 2.8-.4 5.5-.4 8.4c0 53 43 96 96 96s96-43 96-96s-43-96-96-96c-2.8 0-5.6 .1-8.4 .4c5.3 9.3 8.4 20.1 8.4 31.6z"
                fill="currentColor"
              />
            </svg>
            <div
              class="MuiTypography-root MuiTypography-h5 css-hi164r-MuiTypography-root"
            >
              First seen in
            </div>
          </div>
          <div
            class="MuiTypography-root MuiTypography-h5 css-18vl5hq-MuiTypography-root"
          >
            Pilot
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
